First version of detecting minified code.

EnTeQuAk opened the pull request:
Refs #1323

cc @muffinresearch @wagnerand for ideas on what tests I could add to avoid false positives. On my list so far:

 * a file that only has two lines > 500 characters should not match
 * a file with only 8 lines of code but properly indented should not match
 * potentially clone one of our more fully-featured webextension zips and put a modified library in there to avoid any detections of other files in there

Any more ideas?

One more thing I noticed while grepping  through files: We can catch files that have a source map definition in them too, they're very certainly minified. E.g `//# sourceMappingURL=jquery.mobile-1.4.2.min.map` and stuff. ~~I'm working on adding this, that adds less guessing, more knowing to the mix.~~ -> done

muffinresearch commented on src/utils.js:
```diff
@@ -238,3 +237,58 @@ export function parseCspPolicy(policy) {
+
+  // Number of huge lines to also mark a file as potentially minified
+  // Hint: Minified AngularJS has 12 lines, jQuery 4
+  const hugeLinesThreshold = 4;
```
I would consider reducing this to 1? I think it's quite unlikely to have a sane un-minified src file with over 500chars on a single line - or do you have a reason to think that would lead to false-positives?

EnTeQuAk replied:
I could think of base64 encoded files, people do that for absolutely no reason 🤷‍♂️

EnTeQuAk replied:
And, the indentation detection would detect files that are minified but have <4 lines that are super long so I'm not too worried.

muffinresearch commented on src/utils.js:
```diff
@@ -238,3 +237,58 @@ export function parseCspPolicy(policy) {
+  const normalizedCode = code.replace(/\/\*[\S\s]*?\*\/|\/\/.+/g, '');
+
+  while (lines < sampleSize) {
+    lineEndIndex = normalizedCode.indexOf('\n', lineStartIndex);
```
What if a file has no `\n`?

EnTeQuAk replied:
Then it's still matched and could very well be a minified code. Do you mean, if there are absolutely no `\n` we should watch out for longer lines? Or are you referring to hacks like https://github.com/mozilla/addons-server/blob/master/static/js/lib/semicolon.js to fix some compiled code?

muffinresearch replied:
You're going to break out of the loop if no `\n` is found in the entire file. Whilst the check will return true currently you probably want a test to cover this eventually in-case the return logic changes in some way.

EnTeQuAk replied:
I'll add a test

EnTeQuAk replied:
Fixed in e05bb91de236bab06eb62021fb4ed4ea9e700853

muffinresearch commented on src/utils.js:
```diff
@@ -238,3 +237,58 @@ export function parseCspPolicy(policy) {
+  // minified.
+  const indentCountThreshold = 20; // percentage
+
+  // Lenght of a line that looks suspicious of being minified
```
Typo

wagnerand commented:
> a file that only has two lines > 500 characters should not match

Hm, can you elaborate why this should not match?

EnTeQuAk commented:
> Hm, can you elaborate why this should not match?

Well, it might if these are the only two lines and there isn't any other properly indented code but if it's only these two but otherwise indented code it could be just a regular JavaScript file inho.

muffinresearch commented on src/utils.js:
```diff
@@ -238,3 +240,69 @@ export function parseCspPolicy(policy) {
+  }
+
+  return (
+    (indentCount * 100) < indentCountThreshold ||
```
Unless I'm missing something, this doesn't seem to be checking a percentage of checked lines.

EnTeQuAk replied:
indeed, that should be `((indentCount / lines) * 100) < indentCountThreshold`, I blame the keyboard 🙈

EnTeQuAk replied:
Fixed in e3ea083

wagnerand commented:
> Well, it might if these are the only two lines and there isn't any other properly indented code...

But shouldn't this mean the linter should match (ie recognize a minified file)?

EnTeQuAk commented:
> But shouldn't this mean the linter should match (ie recognize a minified file)?

So… let's be a bit more detailed: If a file has exactly two lines that aren't properly indented but <500 chars long it will be detected by our indentation checker since 100% of the lines in the sample size of lines (30) are not indented.

Does this answer your question?

This should be ready for another review now @wagnerand @muffinresearch, I added a few more tests and fixed your comments.

EnTeQuAk commented:
I'm going to merge this so we'll get more testing and throw more code against the detection. if there is anything we can fix them one at a time.
